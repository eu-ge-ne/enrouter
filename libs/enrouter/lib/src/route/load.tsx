import type { ComponentType } from "react";

import { logger } from "#lib/debug.js";

import type { Route } from "./mod.js";

const log = logger("route/load");

export async function loadRoutes(routes: Route[]): Promise<void> {
  log("Loading routes");

  const promises: Promise<unknown>[] = [];

  for (const route of routes) {
    if (route.loaded) {
      continue;
    }

    for (const { fileName, importFn } of route.modules) {
      const promise = loaders[fileName]?.(route, importFn);
      if (promise) {
        promises.push(promise.then(() => (route.loaded = true)));
      }
    }
  }

  await Promise.all(promises);

  log("Routes loaded");
}

type Loader = (
  route: Route,
  importFn: () => Promise<unknown>,
) => Promise<void> | void;

interface _Components {
  components: Record<string, ComponentType>;
}

async function load(importFn: () => Promise<unknown>) {
  const fn = importFn as () => Promise<_Components>;
  const { components } = await fn();
  return Object.fromEntries(
    Object.entries(components).map(([key, C]) => [key, <C />]),
  );
}

export const loaders: Record<string, Loader> = {
  "_layout.tsx": async (route, importFn) => {
    route.elements.layout = await load(importFn);
  },
  "_index.tsx": async (route, importFn) => {
    route.elements.index = await load(importFn);
  },
  "_notFound.tsx": async (route, importFn) => {
    route.elements.notFound = await load(importFn);
  },
};
