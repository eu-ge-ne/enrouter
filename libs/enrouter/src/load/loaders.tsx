import type { ComponentType } from "react";

import type { RouteHandler } from "#handlers/mod.js";

type Loader = (params: {
  handler: RouteHandler;
  module: RouteHandler["modules"][0];
  load: () => Promise<unknown>;
}) => Promise<void> | void;

interface _Components {
  components: Record<string, ComponentType>;
}

async function render(load: () => Promise<unknown>) {
  const fn = load as () => Promise<_Components>;
  const { components } = await fn();
  return Object.fromEntries(
    Object.entries(components).map(([key, C]) => [key, <C />]),
  );
}

export const loaders: Record<string, Loader> = {
  "_layout.tsx": async ({ handler, module, load }) => {
    handler.layout = await render(load);
    module.loaded = true;
  },
  "_index.tsx": async ({ handler, module, load }) => {
    handler.index = await render(load);
    module.loaded = true;
  },
  "_notFound.tsx": async ({ handler, module, load }) => {
    handler.notFound = await render(load);
    module.loaded = true;
  },
};
