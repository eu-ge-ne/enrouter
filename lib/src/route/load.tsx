import type { Route } from "./mod.js";
import { loaders } from "./loaders.js";

export async function loadRoutes(routes: Route[]): Promise<void> {
  const promises: Promise<unknown>[] = [];

  for (const route of routes) {
    if (route.loaded) {
      continue;
    }

    for (const { fileName, importFn } of route.modules) {
      const promise = loaders[fileName]?.(route, importFn);
      if (promise) {
        promises.push(promise.then(() => (route.loaded = true)));
      }
    }
  }

  await Promise.all(promises);
}
