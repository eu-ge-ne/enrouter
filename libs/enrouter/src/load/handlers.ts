import { createLog } from "#log.js";
import { loaders } from "./loaders.js";

import type { RouteModules } from "#modules.js";
import type { RouteHandler } from "#handlers/mod.js";

const log = createLog("load/handlers");

interface LoadRouteHandlersParams {
  handlers: RouteHandler;
  modules: RouteModules;
}

export async function loadRouteHandlers({
  handlers,
  modules,
}: LoadRouteHandlersParams): Promise<void> {
  log("Loading route handlers");

  const promises: Promise<unknown>[] = [];

  recur([handlers], modules, promises);

  await Promise.all(promises);

  log("Route handlers loaded");
}

function recur(
  handlers: RouteHandler[],
  modules: RouteModules,
  promises: Promise<unknown>[],
): void {
  for (const handler of handlers) {
    for (const module of handler.modules) {
      if (module.loaded) {
        continue;
      }

      const { path, load } = modules[module.id]!;

      const fileName = path.slice(path.lastIndexOf("/") + 1);

      const promise = loaders[fileName]?.({ handler, module, load });
      if (promise) {
        promises.push(promise);
      }
    }

    if (handler.tree) {
      recur(handler.tree, modules, promises);
    }
  }
}
