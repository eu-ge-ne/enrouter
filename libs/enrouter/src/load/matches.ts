import { createLog } from "#log.js";

import { loaders } from "./loaders.js";

import type { RouteModules } from "#modules.js";
import type { RouteMatch } from "#matches/mod.js";

const log = createLog("load/matches");

interface LoadRouteMatchesParams {
  matches: RouteMatch[];
  modules: RouteModules;
}

export async function loadRouteMatches({
  matches,
  modules,
}: LoadRouteMatchesParams): Promise<void> {
  log("Loading route matches");

  const promises: Promise<unknown>[] = [];

  for (const { handler } of matches) {
    for (const module of handler.modules) {
      if (module.loaded) {
        continue;
      }

      const { path, load } = modules[module.id]!;

      const fileName = path.slice(path.lastIndexOf("/") + 1);

      const promise = loaders[fileName]?.({ handler, module, load });
      if (promise) {
        promises.push(promise);
      }
    }
  }

  await Promise.all(promises);

  log("Route matches loaded");
}
