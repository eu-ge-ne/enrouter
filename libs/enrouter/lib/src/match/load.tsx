import type { Match } from "./mod.js";
import { loadRoutes } from "#lib/route/load.js";

export function load(matches: Match[]): Promise<void> {
  return loadRoutes(matches.map((x) => x.route).filter((x) => x !== undefined));
}
