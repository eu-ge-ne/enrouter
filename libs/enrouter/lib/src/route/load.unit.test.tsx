import { describe, test, expect } from "vitest";
import * as regexparam from "regexparam";

import { loadRoutes } from "./load.js";

import type { Route } from "./mod.js";

describe("loadRoutes", () => {
  test("no routes", async () => {
    const routes: Route[] = [];

    await loadRoutes(routes);

    expect(routes).toMatchSnapshot();
  });

  test("1 route no modules", async () => {
    const routes: Route[] = [
      {
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: false,
        elements: {},
      },
    ];

    await loadRoutes(routes);

    expect(routes).toMatchSnapshot();
  });

  test("1 route with _layout.tsx module", async () => {
    const routes: Route[] = [
      {
        path: "/",
        test: regexparam.parse("/", true),
        modules: [
          {
            id: "src/_layout.tsx",
            fileName: "_layout.tsx",
            importFn: async () => ({
              components: {
                main: () => <div>Layout</div>,
              },
            }),
          },
        ],
        loaded: false,
        elements: {},
      },
    ];

    await loadRoutes(routes);

    expect(routes).toMatchSnapshot();
  });

  test("1 route with _index.tsx module", async () => {
    const routes: Route[] = [
      {
        path: "/",
        test: regexparam.parse("/", true),
        modules: [
          {
            id: "src/_index.tsx",
            fileName: "_index.tsx",
            importFn: async () => ({
              components: {
                main: () => <div>Index</div>,
              },
            }),
          },
        ],
        loaded: false,
        elements: {},
      },
    ];

    await loadRoutes(routes);

    expect(routes).toMatchSnapshot();
  });
});
