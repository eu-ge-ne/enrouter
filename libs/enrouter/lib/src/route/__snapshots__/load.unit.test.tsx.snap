// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`route > loadRoutes > 0 routes 1`] = `[]`;

exports[`route > loadRoutes > 1 route no modules 1`] = `
[
  {
    "elements": {},
    "loaded": false,
    "modules": [],
    "path": "/",
    "test": {
      "keys": [],
      "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
    },
  },
]
`;

exports[`route > loadRoutes > 1 route with _end.tsx module 1`] = `
[
  {
    "elements": {
      "end": {
        "Main": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {},
          "type": [Function],
        },
      },
    },
    "loaded": true,
    "modules": [
      {
        "fileName": "_end.tsx",
        "id": "src/_end.tsx",
        "importFn": [Function],
      },
    ],
    "path": "/",
    "test": {
      "keys": [],
      "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
    },
  },
]
`;

exports[`route > loadRoutes > 1 route with _index.tsx module 1`] = `
[
  {
    "elements": {
      "index": {
        "Main": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {},
          "type": [Function],
        },
      },
    },
    "loaded": true,
    "modules": [
      {
        "fileName": "_index.tsx",
        "id": "src/_index.tsx",
        "importFn": [Function],
      },
    ],
    "path": "/",
    "test": {
      "keys": [],
      "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
    },
  },
]
`;

exports[`route > loadRoutes > 1 route with _this.tsx module 1`] = `
[
  {
    "elements": {
      "this": {
        "Main": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {},
          "type": [Function],
        },
      },
    },
    "loaded": true,
    "modules": [
      {
        "fileName": "_this.tsx",
        "id": "src/_this.tsx",
        "importFn": [Function],
      },
    ],
    "path": "/",
    "test": {
      "keys": [],
      "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
    },
  },
]
`;
